File: clients/client-js/client-v2/utils/conflation.js
/**
 * Conflation response contract (PLAN_CONFLATION.md §3.3, §4).
 *
 * Conflation — last-value delivery for a consumer group — is declared as a pop
 * query parameter and answered in the pop RESPONSE. Two response keys carry the
 * whole client-side contract, and both are handled here so that pop() and the
 * consume loop cannot drift apart:
 *
 *   "conflation": true          the broker actually applied last-value delivery.
 *                               Rides EVERY conflating response, empty ones
 *                               included — a conflating pop answers 200 with a
 *                               body instead of a bodiless 204 precisely so the
 *                               echo has somewhere to sit.
 *   "conflationConflict": true  this consumer declared a policy the group does
 *                               not have. The STORED group setting wins (§3.3);
 *                               the consumer keeps working and warns once.
 *
 * DEGRADE LOUDLY (§4). No SDK negotiates a version with the broker. A 1.1.0
 * client against an older broker sends `conflation=true`, the broker ignores the
 * unknown query parameter, and the consumer silently processes the entire
 * backlog message by message — the exact silent failure the feature must not
 * have. So: requested conflation + no echo => raise, on the FIRST response,
 * before a single message is handled. `null` (an old broker's 204 on an empty
 * pop) is the most important case to raise on, not the least: it is what an idle
 * queue looks like, and it is the first thing a fresh consumer sees.
 */

import * as logger from './logger.js'

/** `error.code` on the degrade-loudly error, so callers can branch without string-matching. */
export const CONFLATION_UNSUPPORTED = 'conflation_unsupported'

/** The message the plan fixes for every SDK (§4). Keep it identical across clients. */
export const CONFLATION_UNSUPPORTED_MESSAGE =
  'conflation was requested but this broker did not apply it — requires broker >= 1.1.0'

// Once per (queue, group) PER PROCESS, per §3.3: a mismatched fleet during a
// rolling deploy would otherwise emit one warning per pop, forever.
const warnedConflicts = new Set()

// A (queue, group) whose conflict the broker has ALREADY told us about. Proof
// that this broker understands the parameter, so a later response without the
// echo for that same pair is the same known conflict and not an old broker.
//
// It is belt-and-braces now rather than the load-bearing part it once was: a
// 1.1.0 broker keeps the 200 and the body whenever the answer has anything to
// say about conflation (`pop_status`), so an EMPTY conflicting pop carries
// `conflationConflict` too and the check above catches it on the first round
// trip. Keeping the memo costs nothing and still covers a first contact whose
// answer somehow arrives without either key.
const knownConflicts = new Set()

/**
 * Identity of the thing a policy hangs on: the queue (or namespace/task pop
 * target) and the consumer group. Deliberately NOT the affinity key, which
 * includes the partition — the policy is per (queue, group).
 */
export function conflationKey({ queue, namespace, task, group }) {
  const target = queue || `${namespace || '*'}/${task || '*'}`
  return `${target}|${group || '__QUEUE_MODE__'}`
}

/**
 * Inspect one pop response on behalf of a consumer that ASKED for conflation.
 * Call it for every response, including empty ones — that is where the old
 * broker shows itself first.
 *
 * @param {object|null} result - parsed pop response (null for a 204)
 * @param {object} ctx - { queue, namespace, task, group }
 * @throws {Error} with `.code === CONFLATION_UNSUPPORTED` when the broker did
 *   not apply conflation and did not explain why.
 */
export function checkConflationResponse(result, ctx) {
  const key = conflationKey(ctx)

  if (result && result.conflationConflict === true) {
    knownConflicts.add(key)
    if (!warnedConflicts.has(key)) {
      warnedConflicts.add(key)
      logger.warn('Conflation.conflict', {
        queue: ctx.queue || null,
        namespace: ctx.namespace || null,
        task: ctx.task || null,
        group: ctx.group || null,
        message:
          'this consumer declared conflation but the consumer group is registered without it — ' +
          'the STORED group setting wins and this consumer is receiving normal batches. ' +
          'Delete/recreate the group to change its policy.'
      })
    }
    return
  }

  if (result && result.conflation === true) return

  // Pop maintenance: the broker refused the pop before it reached the claim
  // path, so there is no policy to echo and nothing to conclude from the
  // absence of one. Reading it as "old broker" would stop every conflating
  // consumer in the fleet the moment an operator pauses pops.
  if (result && result.paused === true) return

  // The group is known to disagree with us; this response is that same
  // conflict, seen through a body that could not carry the flag.
  if (knownConflicts.has(key)) return

  const error = new Error(CONFLATION_UNSUPPORTED_MESSAGE)
  error.code = CONFLATION_UNSUPPORTED
  throw error
}

/**
 * Test hook: the warn-once and known-conflict registries are process-wide by
 * design, which makes them sticky across tests in one runner process.
 */
export function resetConflationWarnings() {
  warnedConflicts.clear()
  knownConflicts.clear()
}

File: clients/client-py/queen/utils/conflation.py
"""
Client-side half of the conflation contract (PLAN_CONFLATION §3.3, §4).

Two rules live here, and both exist because the broker's answer is the only
channel: there is no capability negotiation anywhere in this SDK.

* **Degrade loudly.** A broker that does not know the `conflation` query
  parameter ignores it and answers an ordinary, successful pop. Nothing raises,
  nothing logs, and the consumer drains the whole backlog one message at a time
  -- the exact failure conflation exists to prevent. So: a response that does
  not carry ``"conflation": true`` after we asked for it is an ERROR, raised on
  the first such response. It is detectable that early because the broker
  emits the key on empty pops too.

* **Warn once on a declaration conflict.** The stored consumer-group policy
  always wins (§3.3): rejecting would break rolling deploys, where half the
  fleet already sends the flag and half does not. But a per-response warning
  would flood at pop rate, so the SDK warns once per (queue, group) per process.

The warn registry is deliberately process-wide and mutable from tests via
``reset_conflict_warnings`` -- "once per process" is not observable otherwise.
"""

from __future__ import annotations

import sys
import threading
from typing import Any, Dict, Optional, Set, Tuple

from . import logger
from ..errors import ConflationUnsupportedError

# The minimum broker that implements PLAN_CONFLATION. Named once so the SDK and
# its error message cannot drift apart.
MIN_BROKER_VERSION = "1.1.0"

# Verbatim across every SDK: operators grep this string, and a per-language
# paraphrase would make that impossible.
UNSUPPORTED_MESSAGE = (
    "conflation was requested but this broker did not apply it "
    "— requires broker >= " + MIN_BROKER_VERSION
)

_conflict_warned: Set[Tuple[str, str]] = set()
_conflict_lock = threading.Lock()


def reset_conflict_warnings() -> None:
    """Forget every (queue, group) that has already warned. Test hook."""
    with _conflict_lock:
        _conflict_warned.clear()


def scope_of(
    queue: Optional[str] = None,
    namespace: Optional[str] = None,
    task: Optional[str] = None,
) -> str:
    """The pop target, as one string, for the warn-once key.

    Mirrors the affinity-key precedence already used for backend routing: a
    named queue when there is one, otherwise namespace:task.
    """
    if queue:
        return queue
    if namespace or task:
        return f"{namespace or '*'}:{task or '*'}"
    return "*"


def warn_conflict_once(queue: Optional[str], group: Optional[str]) -> bool:
    """Emit the declaration-conflict warning at most once per (queue, group).

    Returns True when this call is the one that warned.
    """
    key = (queue or "*", group or "__QUEUE_MODE__")
    with _conflict_lock:
        if key in _conflict_warned:
            return False
        _conflict_warned.add(key)

    logger.warn(
        "Queen.conflation",
        {"queue": key[0], "group": key[1], "status": "declaration-mismatch"},
    )
    # Unconditional, unlike logger.warn (which is off unless QUEEN_CLIENT_LOG=
    # true): a policy this consumer asked for and did not get has to be visible
    # without opting into debug logging. Same channel as the other
    # operator-facing notices in this SDK ("Pop failed", "Network error").
    print(
        f"[queen] conflation conflict on queue '{key[0]}' group '{key[1]}': "
        "the consumer group's stored setting wins and this consumer's "
        "declaration was ignored. Align the declaration, or delete and "
        "recreate the group to change its policy.",
        file=sys.stderr,
    )
    return True


def check_pop_response(
    response: Any,
    *,
    requested: bool,
    queue: Optional[str] = None,
    group: Optional[str] = None,
) -> None:
    """Enforce the two rules above on one pop response.

    Raises ``ConflationUnsupportedError`` when conflation was requested and the
    response says nothing about conflation at all. A response that is not a dict
    (empty body, 204) is treated the same way on purpose: absence of proof is
    not proof of application, and guessing here is what silence costs.

    THE ORDER OF THE CHECKS IS THE CONTRACT, and it is the one every other SDK
    keeps (Go ``checkConflationEcho``, PHP ``ConflationGuard``, JS
    ``checkConflationResponse``, Rust ``check_conflation``): the CONFLICT key is
    read first, because a conflict is a 1.1.0 broker answering "the group is
    already registered the other way, my value wins" -- the opposite of an old
    broker. The body a 1.1.0 broker sends for requested=true / stored=false is
    ``{"messages": [...], "conflationConflict": true}`` with NO ``conflation``
    key (it is emitted only when the effective policy IS conflating, §3.1), so
    raising on the missing echo first would kill exactly the consumer §3.3 Q3
    and §7.3 E2E-4 require to keep running -- and it would do it on that
    consumer's first poll of an idle queue.

    ``paused`` is likewise not a verdict: the broker refuses pops during pop
    maintenance before the request ever reaches the claim path, so there is no
    policy to echo and nothing to conclude from the absence of one.

    A no-op when ``requested`` is falsy -- the check keys off what THIS consumer
    asked for, never off the response alone, so a consumer that never opted in
    is byte-for-byte unaffected.
    """
    if not requested:
        return

    body: Dict[str, Any] = response if isinstance(response, dict) else {}

    if body.get("conflationConflict") is True:
        warn_conflict_once(queue, group)
        return

    if body.get("conflation") is True:
        return

    if body.get("paused") is True:
        return

    raise ConflationUnsupportedError(UNSUPPORTED_MESSAGE)
